packages/server/src/utils/workflow.ts

```diff
@@ -1,127 +1,66 @@
-            // for agent workflow, get the delta from AgentStream event and enqueue it
-            const delta = event.data.delta;
-            if (delta) {
-              controller.enqueue({ delta } as EngineResponse);
-            }
-          } else {
-            appendEventDataToAnnotations(dataStream, event);
-          }
-        }
-      } catch (error) {
-        const errorMessage =
-          error instanceof Error ? error.message : "An unknown error occurred";
-        controller.enqueue({ delta: errorMessage } as EngineResponse);
-        dataStream.close();
-      } finally {
-        controller.close();
-      }
-    },
-  });
-
-  return LlamaIndexAdapter.toDataStreamResponse(stream, {
-    data: dataStream,
-    callbacks: {
-      onFinal: async (content: string) => {
-        const history = chatHistory.concat({ role: "assistant", content });
-        await sendSuggestedQuestionsEvent(dataStream, history);
-        dataStream.close();
-      },
-    },
-  });
-}
-
-async function runCustomWorkflow(
-  workflow: Workflow<AgentWorkflowContext, AgentInputData, string>,
-  agentInput: AgentInputData,
-) {
-  const context = workflow.run(agentInput);
-  const dataStream = new StreamData();
+  let workflowStream:
+    | AsyncIterable<WorkflowEventData<unknown>>
+    | WorkflowStream;
+  if (!input.userInput) {
+    throw new Error("Missing user input to start the workflow");
+  }
+  if (workflow instanceof AgentWorkflow) {
+    workflowStream = workflow.runStream(input.userInput, {
+      chatHistory: input.chatHistory ?? [],
+    });
+  } else {
+    workflowStream = run(workflow, [
+      startAgentEvent.with({
+        userInput: input.userInput,
+        chatHistory: input.chatHistory,
+      }),
+    ]);
+  }
 
-  const stream = new ReadableStream<EngineResponse>({
-    async pull(controller) {
-      try {
-        for await (const event of context) {
-          if (event instanceof StopEvent) {
-            // for normal workflow, the event data from StopEvent is a generator of ChatResponseChunk
-            // iterate over the generator and enqueue the delta of each chunk
-            const generator = event.data as AsyncGenerator<ChatResponseChunk>;
-            for await (const chunk of generator) {
-              controller.enqueue({ delta: chunk.delta } as EngineResponse);
-            }
-          } else {
-            appendEventDataToAnnotations(dataStream, event);
-          }
-        }
-      } catch (error) {
-        const errorMessage =
-          error instanceof Error ? error.message : "An unknown error occurred";
-        controller.enqueue({ delta: errorMessage } as EngineResponse);
-        dataStream.close();
-      } finally {
-        controller.close();
-      }
-    },
-  });
+  const readable = workflowToEngineResponseStream(workflowStream, dataStream);
 
-  return LlamaIndexAdapter.toDataStreamResponse(stream, {
+  return LlamaIndexAdapter.toDataStreamResponse(readable, {
```

Comment:
I think we need here a function that converts the workflow stream directly to a vercel datastream:
```suggestion
  return return new Response(toDataStream(workflowStream), {
    status: 200,
    headers: {
      'Content-Type': 'text/plain; charset=utf-8',
    },
  });
```
`toDataStream` can then be send as a PR to vercel/ai once they merge it, we can directly use it.